### PR TITLE
Option to enforce 2FA on specific users and/or Authentication Providers

### LINKED ISSUE #20: Feature Request: Option to force 2FA for Administrators (or other Editor-Groups) only
It would be awesome to have a way of enforcing 2FA for specified Editor Groups like Administrators.

### PR DESCRIPTION
**Pull request by vcg-development**
We use the OpenIDConnect extension where 2FA is enforced, but we also use the default Username-Password-Provider as a fallback and for all external users, with which we collaberate and which are not in our AD. 

For the latter group we would like to enforce this plugin. Hence, we extended the options to enforce 2FA for not only all users but also for specific users/authentication providers

### CODE REVIEW

**Comment by Benjamin-K** [comment]
I love the new options of this PR. But I would prefer setting a role that is required to set a second factor instead of being required to update the code for every new user. This could even be an abstract role already provided by this package/pr IMO.

**Comment by JamesAlias** [comment]
Thank you for this PR 🙂 

I second the opinion of @Benjamin-K.

- Replace the `enforceTwoFactorAuthenticationForSpecificUserAccounts` option with something like `enforceTwoFactorAuthenticationForRoles`
- Adding a Role to the package that can be used out of the box is not necessary but I'm not against it

File: Classes/Http/Middleware/SecondFactorMiddleware.php
Comment (JamesAlias) [inline Classes/Http/Middleware/SecondFactorMiddleware.php]: Typo here

```diff
-- protected $enforceTwoFactorAuthenticationForAuthenticcationProviders;
++ protected $enforceTwoFactorAuthenticationForAuthenticationProviders;
```

File: Classes/Http/Middleware/SecondFactorMiddleware.php
Comment (JamesAlias) [inline Classes/Http/Middleware/SecondFactorMiddleware.php]: Try to limit unnecessary white space changes

File: Configuration/Settings.yaml
Comment (JamesAlias) [inline Configuration/Settings.yaml]: I'm in favor of expressive variable names, but maybe we can reduce noise here by dropping the `TwoFactorAuthentication` part.

```diff
-- enforceTwoFactorAuthenticationForAllUserAccounts: false
-- enforceTwoFactorAuthenticationForAuthenticationProviders : []
-- enforceTwoFactorAuthenticationForSpecificUserAccounts: []

++ enforceForAllUserAccounts: false
++ enforceForAuthenticationProviders : []
++ enforceForSpecificUserAccounts: []
```

Not necessary and just an idea.

**Review by JamesAlias — CHANGES_REQUESTED** [review]
Some small change requests and ideas.

File: Configuration/Settings.yaml
Comment (Benjamin-K) [reply Configuration/Settings.yaml]: What about `enforce2FAFor...`?

File: Classes/Http/Middleware/SecondFactorMiddleware.php
Comment (vcg-development) [reply Classes/Http/Middleware/SecondFactorMiddleware.php]: Thx

File: Classes/Http/Middleware/SecondFactorMiddleware.php
Comment (vcg-development) [reply Classes/Http/Middleware/SecondFactorMiddleware.php]: Thx, I didn't fire up an IDE on the first commit, hence no csfixer :/

File: Configuration/Settings.yaml
Comment (vcg-development) [reply Configuration/Settings.yaml]: I like the 2FA abbreviation. I reverted the naming of the original setting, so that are no breaking changes. I leave that up to the package maintainer ;)

Comment (vcg-development) [reply Configuration/Settings.yaml]: I also documented the settings wihtin the README.

Thanks for your comments and this awesome package.

Comment (vcg-development) [reply Configuration/Settings.yaml]: Replaced option to verify roles instead of specific users

**Review by JamesAlias — APPROVED** [review]

**Comment by JamesAlias** [comment]
@vcg-development @Benjamin-K 

I have fixed a regression where I could delete the last second factor of an account even though it was enforced by role/authProvider for this account.

I also added a SecondFactorService to centralize the logic and make it more readable on the consuming side.

**Comment by Benjamin-K** [comment]
@JamesAlias Thanks for having a look at this again and cleaning the code up a bit. 

A (new) question raised for me, when reading your last comment: As an administrator I'm not able to remove the last second factor of an account at the moment, if that user is forced to have a second factor, right? 
I think, it should be possible for admins to remove the last second factor of other users no matter what roles they have to help with issues with lost second factors (broken phone, ...). If you agree, I'd create a new issue for that.

**Comment by JamesAlias** [comment]
> @JamesAlias Thanks for having a look at this again and cleaning the code up a bit.
> 
> A (new) question raised for me, when reading your last comment: As an administrator I'm not able to remove the last second factor of an account at the moment, if that user is forced to have a second factor, right? I think, it should be possible for admins to remove the last second factor of other users no matter what roles they have to help with issues with lost second factors (broken phone, ...). If you agree, I'd create a new issue for that.

Good idea. Please do 🙂

**Comment by Benjamin-K** [comment]
Added here: #30

**Pull request merged**
